test/README.md

```diff
@@ -42,26 +42,42 @@ The test suite contains these kind of files:
    ```
    to allow any number of type definitions followed by encoding tests with grammar
    ```
-   <test> = <argseq> <input> <outcome> <desc>
-   <argseq>  ::= ( <argtype>,* )
-   <input> =  <text> | blob <text>
-   <outcome> = true | false
-   <desc> = <text>?
+   <test> ::=
+      | assert <input> : <tuptype> <desc>
+      | assert <input> !: <tuptype> <desc>
+      | assert <input> == <input> : <tuptype> <desc>
+      | assert <input> != <input> : <tuptype> <desc>
```

Comment:
Also assert subtyping? `assert <tuptype> <: <tuptype> <desc>`

Reply:
Potentially, although many Candid implementations don't implement subtyping as a relation on types, but only implicitly as “this value can be parsed at this type”. So
```
assert "DIDL\00\02\7e\7f\01)" : (bool)
```
is actually testing subtyping the subtyping of `(bool, null) <: (bool)`.

So I don’t see an urgent need, but we can of course extend the syntax (maybe later, to iterate in smaller steps)?

Reply:
> Potentially, although many Candid implementations don't implement subtyping as a relation on types, but only implicitly as “this value can be parsed at this type”

I came across a similar dissonance for `candiff` when it accepts and checks the value against a given type.

Does this check permit a subtyping relation?  A reasonable user may wonder this.

Currently it does not, and @chenyan-dfinity explained why with a similar point (or so it seemed to me).

In that sense, it's overly conservative about what an _implementation of the service at this type_ would do, right?  IIUC, that's the case, and we just don't worry for now?

My concern is that our CLI tools, which are meant to inform, will be confusing on this point if we don't resolve it there, at least.  (Imagine a developer using these tools to predict what candid values can and cannot be sent at a given type, a reasonable use case.)

Reply:
No, it does support subtyping, e.g. `(42, bool) : (nat) == (42)`. What's not implemented yet is the explicit subtyping relation, i.e. the `<:` relation over types. It's not required for a serialization library, but we do need to support subtyping when encoding/decoding value at a given type.